### PR TITLE
Don't sort postings if we only have one block.

### PR DESCRIPTION
**Pull request by brian-brazil**
This will make big queries that only touch the head faster,
though queries that touch both the head and a block will still
be the same speed. This probably won't help much with graphing
unless the range is under an hour, however it should make most
recording rules faster. 

PromQL benchmarks for histograms show only 2-3% improvement, but
they're only over 1k series. Looks like we'd be saving about 1s for a head-only
query that touches 1M series, so a 1M series histogram_quantile(rate(5m)) 
would go from around 12s to 11s.

### CODE REVIEW

File: tsdb/querier.go [hunk 1]
```diff
@@ -106,14 +109,29 @@ func (q *querier) lvals(qs []Querier, n string) ([]string, error) {
...
+func (q *querier) SelectSorted(ms ...*labels.Matcher) (SeriesSet, error) {
 	if len(q.blocks) == 0 {
 		return EmptySeriesSet(), nil
 	}
 	ss := make([]SeriesSet, len(q.blocks))
 	var s SeriesSet
 	var err error
 	for i, b := range q.blocks {
-		s, err = b.Select(ms...)
+		if len(q.blocks) == 1 {
```
Comment (roidelapluie) [inline tsdb/querier.go]: When we call this function and q.blocs is 1 ?

File: tsdb/querier.go [hunk 2]
```diff
@@ -217,6 +239,24 @@ func (q *blockQuerier) Select(ms ...*labels.Matcher) (SeriesSet, error) {
 	}, nil
 }
 
+func (q *blockQuerier) SelectSorted(ms ...*labels.Matcher) (SeriesSet, error) {
```
Comment (roidelapluie) [inline tsdb/querier.go]: What is the difference between this and Select ?

Comment (roidelapluie) [reply tsdb/querier.go]: nevermind I have seent he difference

File: tsdb/querier.go [hunk 3]
```diff
@@ -106,14 +109,29 @@ func (q *querier) lvals(qs []Querier, n string) ([]string, error) {
...
 func (q *querier) Select(ms ...*labels.Matcher) (SeriesSet, error) {
+	if len(q.blocks) != 1 {
+		return q.SelectSorted(ms...)
+	}
+	// Sorting Head series is slow, and unneeded when only the
+	// Head is being queried. Sorting blocks is a noop.
+	return q.blocks[0].Select(ms...)
+}
+
+func (q *querier) SelectSorted(ms ...*labels.Matcher) (SeriesSet, error) {
```
Comment (roidelapluie) [inline tsdb/querier.go]: do we need this publicly?

**Comment by roidelapluie** [comment]
This optimization look like implementation details that we could not expose publicly.

File: tsdb/querier.go (hunk 3 above)
Comment (brian-brazil) [reply tsdb/querier.go]: It's the Query interface that we're using everywhere, having shadow internal interfaces that are slightly different tends to end up being hard to maintain

File: tsdb/querier.go (hunk 1 above)
Comment (brian-brazil) [reply tsdb/querier.go]: Ah, that's from when I refactored this a bit. Let me fix that.

Comment (roidelapluie) [reply tsdb/querier.go]: if the function is exposed publicly that might still make sense

**Comment by brian-brazil** [comment]
> This optimization look like implementation details that we could not expose publicly.

That's the nature of some optimisations, they require removing a level of abstraction.

File: tsdb/querier.go [hunk 4]
```diff
@@ -106,14 +109,29 @@ func (q *querier) lvals(qs []Querier, n string) ([]string, error) {
...
 	if len(q.blocks) == 0 {
 		return EmptySeriesSet(), nil
 	}
```
Comment (roidelapluie) [inline tsdb/querier.go]: I feek like we could avoid this extra function if we pu the if len ==  1 there and return q.blocks[0].Select(ms...)

File: tsdb/querier.go (hunk 1 above)
Comment (brian-brazil) [reply tsdb/querier.go]: If you call the Sorted version of the function, the result should still be sorted.

File: tsdb/querier.go (hunk 4 above)
Comment (brian-brazil) [reply tsdb/querier.go]: Not quite, as the sorting has to happen another function down again.

File: tsdb/querier.go (hunk 1 above)
Comment (roidelapluie) [reply tsdb/querier.go]: As we just have one block arent the postings sorted already?

File: tsdb/querier.go (hunk 4 above)
Comment (roidelapluie) [reply tsdb/querier.go]: As we just have one block arent the postings sorted already?

Comment (roidelapluie) [reply tsdb/querier.go]: mmh neverming in the head block it is probaly not sorted

Comment (brian-brazil) [reply tsdb/querier.go]: Exactly

**Comment by roidelapluie** [comment]
Merge conflict + waiting for another tsdb contributor to review this

File: tsdb/querier.go (hunk 3 above)
Comment (bwplotka) [reply tsdb/querier.go]: Yes please (:

**Comment by brian-brazil** [comment]
> This is because all the GO users of the TSDB who imports code will be surprised

We never guaranteed any particular sorting, and they'd lose the performance benefits.

**Comment by bwplotka** [comment]
Hm, yea I thought it was guaranteed somehow and e.g promql rely on this to produce certain order of results, but looks like we did not guaranteed that indeed, so LGTM (:

**Comment by bwplotka** [comment]
What about remote read did we not guarantee the order of returned series? And it bases on Select. @brian-brazil

**Comment by brian-brazil** [comment]
We have grouping for remote read, but not ordering.

**Comment by brian-brazil** [comment]
Hmm, that's not part of the protocol (and not really sane to have as part of it as the sort order is entirely internal to the tsdb) so that sounds like we need to make the remote read code more robust.

**Comment by bwplotka** [comment]
Let's figure out tomorrow how we can solve this (:

**Comment by brian-brazil** [comment]
@bwplotka updated as discussed.

**Comment by brian-brazil** [comment]
Added.

File: tsdb/querier_bench_test.go [hunk 5]
```diff
@@ -136,7 +136,11 @@ func BenchmarkQuerierSelect(b *testing.B) {
 	}
 	testutil.Ok(b, app.Commit())
 
-	bench := func(b *testing.B, br BlockReader) {
+	emptyHead, err := NewHead(nil, nil, nil, 1000)
+	testutil.Ok(b, err)
+	defer emptyHead.Close()
```
Comment (codesome) [inline tsdb/querier_bench_test.go]: I don't see this being used anywhere

File: web/api/v1/api.go [hunk 6]
```diff
@@ -1121,7 +1121,12 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 		for i, query := range req.Queries {
-			err := api.remoteReadQuery(ctx, query, externalLabels, func(set storage.SeriesSet) error {
+			err := api.remoteReadQuery(ctx, query, externalLabels, func(querier storage.Querier, selectParams *storage.SelectParams, filteredMatchers []*labels.Matcher) error {
+				// The streaming API provides sorted series.
+				set, _, err := querier.SelectSorted(selectParams, filteredMatchers...)
```
Comment (codesome) [inline web/api/v1/api.go]: Why is it only the streaming API that has it sorted?

Comment (brian-brazil) [reply web/api/v1/api.go]: Thanos needs it. I don't see a reason to slow down the older endpoint, nor to incorrectly make it seem that it must be sorted.

File: tsdb/querier_bench_test.go (hunk 5 above)
Comment (brian-brazil) [reply tsdb/querier_bench_test.go]: Fixed

File: web/api/v1/api.go (hunk 6 above)
Comment (codesome) [reply web/api/v1/api.go]: :+1:

Comment (bwplotka) [reply web/api/v1/api.go]: Yes, we rely on this heavily. Thanks!

**Pull request merged**